### ISSUE TITLE
[strict-boolean-expressions] unexpected error inside HOC

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#4877]
### Bug Report

- __TSLint version__: 5.20.
- __TypeScript version__: 3.6.3
- __Running TSLint via__: CLI

#### Reproduction using [TSLint Playground](https://palantir.github.io/tslint-playground)

(Requires React types so it wasn't possible to create a playground link.)

#### TypeScript code being linted

```ts
import React, { Component } from 'react';

type ShouldDetectVisibilityProp = {
  shouldDetectVisibility: boolean;
};

const withIsVisible = <ComposedProps,>(ComposedComponent: React.ComponentClass<ComposedProps>) => {
  type Props = ComposedProps & ShouldDetectVisibilityProp;

  class ComponentWithIsVisible extends Component<Props> {
    render() {
      // Unexpected error: This type is not allowed in the condition because it is always truthy. Only booleans are allowed.
      return this.props.shouldDetectVisibility ? 'yes' : 'no';
    }
  }

  return ComponentWithIsVisible;
};
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-boolean-expressions": [true]
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
